diff --git a/EntityManager.d.ts b/EntityManager.d.ts
index 06c4103793827966d8addba05e9b7459199f2cd5..a82282335f6812d16f6d0db730ab8cb36924ce1d 100644
--- a/EntityManager.d.ts
+++ b/EntityManager.d.ts
@@ -577,8 +577,6 @@ export interface MergeOptions {
     refresh?: boolean;
     convertCustomTypes?: boolean;
     schema?: string;
-    disableContextResolution?: boolean;
-    keepIdentity?: boolean;
 }
 export interface ForkOptions {
     /** do we want a clear identity map? defaults to true */
diff --git a/EntityManager.js b/EntityManager.js
index 12803d901d230b66ba388fe4dfa769bf14f7db33..c12f5f710a172ad6853a19994e93e64d5fdeb80b 100644
--- a/EntityManager.js
+++ b/EntityManager.js
@@ -1181,10 +1181,10 @@ class EntityManager {
      * via second parameter. By default, it will return already loaded entities without modifying them.
      */
     merge(entityName, data, options = {}) {
+        const em = this.getContext();
         if (utils_1.Utils.isEntity(entityName)) {
-            return this.merge(entityName.constructor.name, entityName, data);
+            return em.merge(entityName.constructor.name, entityName, data);
         }
-        const em = options.disableContextResolution ? this : this.getContext();
         options.schema ??= em._schema;
         entityName = utils_1.Utils.className(entityName);
         em.validator.validatePrimaryKey(data, em.metadata.get(entityName));
@@ -1194,12 +1194,7 @@ class EntityManager {
         }
         const meta = em.metadata.find(entityName);
         const childMeta = em.metadata.getByDiscriminatorColumn(meta, data);
-        const dataIsEntity = utils_1.Utils.isEntity(data);
-        if (options.keepIdentity && entity && dataIsEntity && entity !== data) {
-            em.entityFactory.mergeData(meta, entity, (0, entity_1.helper)(data).__originalEntityData, { initialized: true, merge: true, ...options });
-            return entity;
-        }
-        entity = dataIsEntity ? data : em.entityFactory.create(entityName, data, { merge: true, ...options });
+        entity = utils_1.Utils.isEntity(data) ? data : em.entityFactory.create(entityName, data, { merge: true, ...options });
         em.validator.validate(entity, data, childMeta ?? meta);
         em.unitOfWork.merge(entity);
         return entity;
diff --git a/utils/TransactionManager.js b/utils/TransactionManager.js
index 8198287fe2bc33c6e81bf4efd5938cf56d823924..4dc559d0515f728f6748862ead5387bb8a3a4e21 100644
--- a/utils/TransactionManager.js
+++ b/utils/TransactionManager.js
@@ -153,7 +153,8 @@ class TransactionManager {
      */
     mergeEntitiesToParent(fork, parent) {
         for (const entity of fork.getUnitOfWork(false).getIdentityMap()) {
-            parent.merge(entity, { disableContextResolution: true, keepIdentity: true, refresh: true });
+            parent.unitOfWork.register(entity);
+            entity.__helper.__em = parent;
         }
     }
     /**
